src/storage/UpdateVertexProcessor.cpp

```diff
@@ -0,0 +1,360 @@
+                                            nebula::cpp2::SupportedType::INT);
+                   break;
+               }
+               case VAR_DOUBLE: {
+                   writer << boost::get<double>(v);
+                   column = this->columnDef(std::string("anonymous"),
+                                            nebula::cpp2::SupportedType::DOUBLE);
+                   break;
+               }
+               case VAR_BOOL: {
+                   writer << boost::get<bool>(v);
+                   column = this->columnDef(std::string("anonymous"),
+                                            nebula::cpp2::SupportedType::BOOL);
+                   break;
+               }
+               case VAR_STR: {
+                   writer << boost::get<std::string>(v);
+                   column = this->columnDef(std::string("anonymous"),
+                                            nebula::cpp2::SupportedType::STRING);
+                   break;
+               }
+               default: {
+                   LOG(FATAL) << "Unknown VariantType: " << v.which();
+                   return;
+               }
+           }
+           respScheam.columns.emplace_back(std::move(column));
+        }
+        resp_.set_schema(std::move(respScheam));
+        resp_.set_data(std::move(writer.encode()));
+    }
+}
+
+
+kvstore::ResultCode UpdateVertexProcessor::collectVertexProps(
+                            const PartitionID partId,
+                            const VertexID vId,
+                            const TagID tagId,
+                            const std::vector<PropContext>& props) {
+    auto prefix = NebulaKeyUtils::prefix(partId, vId, tagId);
+    std::unique_ptr<kvstore::KVIterator> iter;
+    auto ret = this->kvstore_->prefix(this->spaceId_, partId, prefix, &iter);
+    if (ret != kvstore::ResultCode::SUCCEEDED) {
+        VLOG(3) << "Error! ret = " << static_cast<int32_t>(ret)
+                << ", spaceId " << this->spaceId_;
+        return ret;
+    }
+
+    if (iter && iter->valid()) {
+        auto reader = RowReader::getTagPropReader(this->schemaMan_,
+                                                  iter->val(),
+                                                  this->spaceId_,
+                                                  tagId);
+        const auto constSchema = reader->getSchema();
+        for (auto prop : props) {
+            auto index =  constSchema->getFieldIndex(prop.prop_.name);
+            auto res = RowReader::getPropByIndex(reader.get(), index);
+            if (!ok(res)) {
+                VLOG(1) << "Skip the bad value for tag: " << tagId
+                        << ", prop " << prop.prop_.name;
+                return kvstore::ResultCode::ERR_UNKNOWN;
+            }
+            auto&& v = value(std::move(res));
+            tagFilters_.emplace(std::make_pair(tagId, prop.prop_.name), v);
+        }
+        if (updateTagIds_.find(tagId) != updateTagIds_.end()) {
```

Comment:
What's the updateTagIds_ used for?

Reply:
I guess, for each tagId  need to update or upsert, we get corresponding schema and updater, if it pass filter, then we can update it.

Reply:
@critical27 thx, you are all right.